### ISSUE TITLE
Streams javadsl does not support Java variance

### ISSUE THREAD

**Issue by jroper** [akka/akka#17806]
All the methods in the Java stream DSL appear to be invariant.  The function types they take are contravariant on the argument and covariant on the return type, but the Java compiler doesn't see or support that, it only supports variance at the call site, so variance must be declared by the methods that take parameters.  For example, let's say you had defined a generic application wide `akka.japi.Function` for converting numbers to Strings in a domain specific way, and then you had a flow of integers, the following won't compile:

```java
akka.japi.Function<Number, String> numberToString = getNumberToStringFunction();
akka.streams.javadsl.Flow<Integer, Integer, ?> flow = createFlow();
flow.map(numberToString);
```

Because `Flow.map` is invariant on the input type of Function.

**Comment by drewhk** [comment]
Very true. Added to RC4.

**Comment by drewhk** [comment]
Moved to post 1.0 due to time constraints

**Comment by rkuhn** [comment]
This is fixed in #18605.
